### ISSUE TITLE
watchStream.Cancel blocks when it should not

### ISSUE THREAD

**Issue by nlacasse** [vanadium/issues#1284]
According to the [documentation of Stream](https://godoc.org/v.io/v23/syncbase#Stream): `Cancel does not block.`  This behavior is necessary so that cancel `can  be called concurrently with a goroutine that is iterating via Advance.`

However, our implementation of `WatchStream.Cancel`  blocks.  It waits on a Lock that is also held by Advance(), and there is a TODO to fix it.

From v.io/v23/syncbase/watch_stream.go:

```
// Cancel implements Stream interface.
// TODO(rogulenko): Make Cancel non-blocking.
func (s *watchStream) Cancel() {
    s.mu.Lock()
    defer s.mu.Unlock()
    s.cancel()
    s.err = verror.New(verror.ErrCanceled, nil)
}
```

A similar lock-holding exists in v.io/x/ref/services/syncbase/watchable/stream.go:

```
// Cancel implements the store.Stream interface.
func (s *stream) Cancel() {
    s.mu.Lock()
    defer s.mu.Unlock()
    if s.err != nil {
        return
    }
    s.iit.Cancel()
}
```

**Comment by ivanpi-mtv** [comment]
Fixed client-visible blocking in v.io/c/21604 . Need to investigate backend, lower priority since it is isolated from the client through RPC cancellation.
